### PR TITLE
Add a dedicated type for issues

### PR DESCRIPTION
**Pull request by neithernut**
Blocks #99.

### CODE REVIEW

**Comment by neithernut** [comment]
Mh... originally I planned on creating the new `Issue` type in the root module of the library and moving the repository over, too. Now I'm not so sure any more whether having the `Issue` in the root module is such a good idea. Issues are virtually always queried from the repository anyway so the complete module path may not show up as often as I anticipated.
Also, an `issue` could also contain some issue-specific helper types (in the same file).
Any thoughts, @matthiasbeyer?

**Comment by neithernut** [comment]
Moving the `Issue` type into a new module had the effect that `Issue::new` has to be public for `repository::find_issue()` to compile. I would have liked to not expose that function.

File: lib/src/issue.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        self.id
+    }
+
+    /// Get possible heads of the issue
+    ///
+    /// Returns the head references from both the local repository and remotes
+    /// for this issue.
+    ///
+    pub fn heads(&self) -> Result<References<'r>> {
+        let glob = format!("**/dit/{}/head", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Don't use `self` here for formatting! If the impl of `Display` changes, this will result in strange errors which are hard to debug. Explicit > Implicit!

File: lib/src/issue.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get the local issue head for the issue
+    ///
+    /// Returns the head reference of the issue from the local repository, if
+    /// present.
+    ///
+    pub fn find_local_head(&self) -> Result<Reference<'r>> {
+        let refname = format!("refs/dit/{}/head", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: Same here.

File: lib/src/issue.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get the leaf references for the issue
+    ///
+    /// Returns the leaf references for the issue from both the local repository
+    /// and remotes.
+    ///
+    pub fn issue_leaves(&self) -> Result<References<'r>> {
+        let glob = format!("**/dit/{}/leaves/*", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: And here...

File: lib/src/issue.rs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get all local references for the issue
+    ///
+    /// Return all references associated with the issue from the local
+    /// repository.
+    ///
+    pub fn local_refs(&self) -> Result<References<'r>> {
+        let glob = format!("refs/dit/{}/**", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: And here.

File: lib/src/issue.rs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            .references_glob(&glob)
+            .chain_err(|| EK::WrappedGitError)
+    }
+
+    /// Get a revwalk for traversing all messages of the issue
+    ///
+    /// The sorting of the revwalk will be set to "topological".
+    ///
+    pub fn message_revwalk(&self) -> Result<git2::Revwalk<'r>> {
+        let glob = format!("**/dit/{}/**", self);
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: And here.

File: src/main.rs [hunk 6]
```diff
@@ -412,7 +416,7 @@ fn show_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
         if matches.is_present("initial") {
             vec![(TreeGraphElemLine::empty(), try_or_1!(repo.find_commit(issue)))]
         } else {
-            try_or_1!(repo.get_issue_revwalk(issue))
+            try_or_1!(repo.find_issue(issue).and_then(|issue| issue.message_revwalk()))
```
Comment (matthiasbeyer) [inline src/main.rs]: You can put the `and_then()` call on a new line here.

Also (I'm not sure with this one) you could use `and_then(Issue::message_revwalk)` which would result in a bit cleaner code.

File: src/main.rs [hunk 7]
```diff
@@ -463,7 +467,7 @@ fn tag_impl(repo: &Repository, matches: &clap::ArgMatches) -> i32 {
     let issue = try_or_1!(Oid::from_str(matches.value_of("issue-hash").unwrap()));
 
     // get the head for the issue to tag
-    let mut issue_head = try_or_1!(repo.get_local_issue_head(issue));
+    let mut issue_head = try_or_1!(repo.find_issue(issue).and_then(|issue| issue.find_local_head()));
```
Comment (matthiasbeyer) [inline src/main.rs]: Same points as above (the first one only if you care, because everything fits in one line, still).

File: lib/src/issue.rs (hunk 1 above)
Comment (neithernut) [inline lib/src/issue.rs]: It is meant to do just that. The rationale is that the implementation of `Display` always results in a unique string which may be used as part of reference names. This way, if we introduce "scoping" or "namespacing" for issues, the scope or namespace finds its way into references automatically.

We also could have a function which explicitly returns a string referring to the issue, which is specified to be used in ref-names and have the `Display` implementation rely on that function instead. If you insist, that is.

File: src/main.rs (hunk 6 above)
Comment (neithernut) [inline src/main.rs]: Not quite. You have to get a reference first.

File: lib/src/issue.rs (hunk 1 above)
Comment (matthiasbeyer) [inline lib/src/issue.rs]: The problem I'm talking about is that `Display` is explicitely for returning a string which gets shown to the user:

> Display is similar to Debug, but Display is for user-facing output

(from [here](https://doc.rust-lang.org/std/fmt/trait.Display.html))

Therefor, it's implementation may change to return something like

```
"Issue d630a948079bee0adc5f"
```

which would result in huge bugs.

It is not intended to return an identifier or something like that.

**Comment by neithernut** [comment]
I squashed and rebased on the new master, with new error kinds.

File: lib/src/issue.rs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            .and_then(|mut revwalk| {
+                // The iterator will iterate over all the messages in the tree
+                // spanned but it will halt at the initial message.
+                revwalk.push_glob(glob.as_ref())?;
+                self.repo
+                    .find_commit(self.id)
+                    .and_then(|commit| commit.parent_id(0))
+                    .ok() // the initial message having no parent is not unusual
+                    .map(|parent| revwalk.hide(parent))
+                    .unwrap_or(Ok(()))?;
```
Comment (matthiasbeyer) [inline lib/src/issue.rs]: I find that line a bit hard to follow (as in "why a default value here... we do not use it"). That would be clearer if we'd use a `let _ = self.repo` at the beginnging of this block of code... just a readability thing!

File: src/main.rs (hunk 6 above)
Comment (matthiasbeyer) [inline src/main.rs]: Either way you can put the `and_then` on a new line here.

File: src/main.rs (hunk 7 above)
Comment (matthiasbeyer) [inline src/main.rs]: ...and here.
